tests/core/object/test_undo_redo.h

```diff
@@ -0,0 +1,85 @@
+#ifndef TEST_UNDO_REDO_H
+#define TEST_UNDO_REDO_H
+
+#include "core/object/undo_redo.h"
+#include "tests/test_macros.h"
+
+// Declared in global namespace because of GDCLASS macro warning (Windows):
+// "Unqualified friend declaration referring to type outside of the nearest enclosing namespace
+// is a Microsoft extension; add a nested name specifier".
+class _TestUndoRedoObject : public Object {
+	GDCLASS(_TestUndoRedoObject, Object);
+	int property_value = 0;
+
+protected:
+	static void _bind_methods() {
+		ClassDB::bind_method(D_METHOD("set_property", "property"), &_TestUndoRedoObject::set_property);
+		ClassDB::bind_method(D_METHOD("get_property"), &_TestUndoRedoObject::get_property);
+		ADD_PROPERTY(PropertyInfo(Variant::INT, "property"), "set_property", "get_property");
+	}
+
+public:
+	void set_property(int value) { property_value = value; }
+	int get_property() const { return property_value; }
+	void increment_property(int value) { property_value = value; }
+};
+
+namespace TestUndoRedo {
+
+void set_property_action(UndoRedo *undo_redo, const String &name, _TestUndoRedoObject *test_object, int value, UndoRedo::MergeMode merge_mode = UndoRedo::MERGE_DISABLE) {
+	undo_redo->create_action("Set Property", merge_mode);
+	undo_redo->add_do_property(test_object, "property", value);
+	undo_redo->add_undo_property(test_object, "property", test_object->get_property());
+	undo_redo->commit_action();
+}
+
+TEST_CASE("[UndoRedo] Simple Property UndoRedo") {
+	GDREGISTER_CLASS(_TestUndoRedoObject);
+	UndoRedo undo_redo = UndoRedo();
+
+	_TestUndoRedoObject test_object = _TestUndoRedoObject();
+
+	CHECK(test_object.get_property() == 0);
+	CHECK(undo_redo.get_version() == 1);
+	CHECK(undo_redo.get_history_count() == 0);
+
+	set_property_action(&undo_redo, "Set Property", &test_object, 10);
+
+	CHECK(test_object.get_property() == 10);
+	CHECK(undo_redo.get_version() == 2);
+	CHECK(undo_redo.get_history_count() == 1);
+
+	undo_redo.undo();
+
+	CHECK(test_object.get_property() == 0);
+	CHECK(undo_redo.get_version() == 1);
+	CHECK(undo_redo.get_history_count() == 1);
+
+	undo_redo.redo();
+
+	CHECK(test_object.get_property() == 10);
+	CHECK(undo_redo.get_version() == 2);
+	CHECK(undo_redo.get_history_count() == 1);
+
+	set_property_action(&undo_redo, "Set Property", &test_object, 100);
+
+	CHECK(test_object.get_property() == 100);
+	CHECK(undo_redo.get_version() == 3);
+	CHECK(undo_redo.get_history_count() == 2);
+
+	set_property_action(&undo_redo, "Set Property", &test_object, 1000);
+
+	CHECK(test_object.get_property() == 1000);
+	CHECK(undo_redo.get_version() == 4);
+	CHECK(undo_redo.get_history_count() == 3);
+
+	undo_redo.undo();
+
+	CHECK(test_object.get_property() == 100);
+	CHECK(undo_redo.get_version() == 3);
+	CHECK(undo_redo.get_history_count() == 3);
+}
+
+} //namespace TestUndoRedo
+
+#endif // TEST_UNDO_REDO_H
```

Comment:
```suggestion
#endif // TEST_UNDO_REDO_H

```
Missing line termination

Reply:
Weird, the extra line at the end is definitely in my VSCode editor, and clang-format sets it that way... but when I push it it doesn't seem to be in the commit

Reply:
Done